clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
 
     return eraseAnonNamespace(ArgOS.str().str());
   }
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
```

Comment:
Does this work if it is a type defined in a lambda?  Something like:

    namespace NS {
      struct Foo {
          void bar() {
              []() {
                   struct Here{void operator()();};
                    }
        }
      };
    }

The type 'Here' is named something like: NS::Foo::bar()::{lambda()#1}::operator()() const::Here.  

Though, actually, I'm not sure how to forward declare that type... this might just be a general bug (kernels with names in an anonymous type).  @mkinsner @rolandschulz  what does the standard say about that?